### ISSUE TITLE
Test is assigned to undefined emulator on second run

### ISSUE THREAD

**Issue by vjmaiky** [wix/Detox#2320]
**Describe the bug**
I followed all the steps from documentation (which worked smooth together with a mock rn project done by me):
https://github.com/wix/Detox/blob/master/docs/Introduction.Android.md
Everything works ok until i try to run 2 sets of 1 test each one by one, or the same tests twice.
First test it is executed correctly (and it succeed), but when second set starts, i get this:

```
detox[84720] DEBUG: [AsyncWebSocket.js/WEBSOCKET_OPEN] opened web socket to: ws://localhost:58740
detox[84720] DEBUG: [DetoxServer.js/LOGIN] role=tester, sessionId=c860fc25-0231-e36f-4bb1-71a8db3e33aa
detox[84720] DEBUG: [DetoxServer.js/LOGIN_SUCCESS] role=tester, sessionId=c860fc25-0231-e36f-4bb1-71a8db3e33aa
detox[84720] DEBUG: [exec.js/EXEC_CMD, #0] "/Users/admin/Library/Android/sdk/emulator/emulator" -list-avds --verbose
detox[84720] DEBUG: [exec.js/EXEC_CMD, #1] "/Users/admin/Library/Android/sdk/emulator/emulator" -version
detox[84720] DEBUG: [EmulatorVersionResolver.js/EMU_BIN_VERSION_DETECT] Detected emulator binary version { major: 30, minor: 0, patch: 26 }
detox[84720] DEBUG: [AndroidDriver.js/ALLOCATE_DEVICE] Trying to allocate a device based on "Pixel_API_28_AOSP"
detox[84720] DEBUG: [exec.js/EXEC_CMD, #2] "/Users/admin/Library/Android/sdk/platform-tools/adb"  devices
detox[84720] DEBUG: [exec.js/EXEC_SUCCESS, #2] List of devices attached
emulator-15636  device


detox[84720] ERROR: [EmulatorTelnet.js/TELNET_TIMEOUT] 
detox[84720] ERROR: Error: response not received
detox[84720] INFO:  Second fixture is assigned to undefined
detox[84720] INFO:  Second fixture: running 1st test
detox[84720] INFO:  Second fixture: running 1st test [SKIPPED]

```
**To Reproduce**
First run of a test (successful): https://gist.github.com/vjmaiky/ce70ba0fe0fc4dec389897b37ee091e7
Second run of the same test (failing): https://gist.github.com/vjmaiky/1ea83ae7f1d377e3f7e2f5579803e749
To be able to run again the test i need to manually close the emulator.
On ios i do not have this issue. There i have about 10 sets with more than 2 tests each, and working smooth.

**Environment (please complete the following information):**
Dependencies:
```
 - Detox:17.4.7
 - React Native:0.63.2
 - Node:v12.4.0
 - Device: Android Emulator
 - OS: API 28
 - Test-runner (select one): `jest-circus` 26.4.2

**Comment by vjmaiky** [comment]
After trying to find some extra infos i saw someone complaining about something similar, and his resolution was to downgrade RN to 0.62. Now i have also RN 0.62.2 but still not working; more details here:
https://stackoverflow.com/questions/64046997/android-test-fails-on-second-run-if-emulator-is-already-running

Any hints?

**Comment by mrbrentkelly** [comment]
Getting something similar which has just started randomly happening yesterday on my local Mac as well as on TeamCity (which is also using Mac's).

Basically my tests run fine but after the last test completes the current run hangs indefinitely with the following error...

```
detox[8102] ERROR: [EmulatorTelnet.js/TELNET_TIMEOUT] 
detox[8102] ERROR: Error: response not received
 PASS  e2e/tests/MyComponent.e2e.ts (36.297s)
  MyComponent E2E tests
    ✓ performs a successful log in using username and password (7691ms)

detox[8102] WARN:  at node_modules/jest-cli/build/cli/index.js:271:15 
 Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
``` 

I've also noticed that if my Android emulator is already running when I initiate the detox tests it fails almost immediately...

```
$ detox test --configuration android
detox[9084] INFO:  [test.js] configuration="android" useCustomLogger=true forceAdbInstall=false DETOX_START_TIMESTAMP=1602751411073 readOnlyEmu=false reportSpecs=true jest --config e2e/config.json --testNamePattern '^((?!:ios:).)*$' --maxWorkers 1 e2e
detox[9085] INFO:  [DetoxServer.js] server listening on localhost:52716...
detox[9085] ERROR: Error: response not received
detox[9085] INFO:  MyComponent E2E tests is assigned to undefined
detox[9085] INFO:  MyComponentn E2E tests: performs a successful log in using username and password
detox[9085] INFO:  MyComponent E2E tests: performs a successful log in using username and password [SKIPPED]

detox[9085] ERROR: [EmulatorTelnet.js/TELNET_TIMEOUT] 
detox[9085] ERROR: [EmulatorTelnet.js/TELNET_ERROR] connect ECONNREFUSED 127.0.0.1:23
  err: Error: connect ECONNREFUSED 127.0.0.1:23
      at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1141:16)
detox[9085] ERROR: Error: connect ECONNREFUSED 127.0.0.1:23
detox[9085] WARN:  at node_modules/jest-cli/build/cli/index.js:271:15 
 Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
```

I've tried updating to the latest version of Detox (17.9.0) but the issue still persists.

Here's my build info:

```
info Fetching system and libraries information...
System:
    OS: macOS 10.15.7
    CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
    Memory: 859.38 MB / 32.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 12.16.3 - ~/.nvm/versions/node/v12.16.3/bin/node
    Yarn: 1.22.4 - ~/.nvm/versions/node/v12.16.3/bin/yarn
    npm: 6.14.4 - ~/.nvm/versions/node/v12.16.3/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.9.3 - /Users/brent.kelly/.rvm/gems/ruby-2.6.3/bin/pod
  SDKs:
    iOS SDK:
      Platforms: iOS 14.0, DriverKit 19.0, macOS 10.15, tvOS 14.0, watchOS 7.0
    Android SDK:
      Android Debug Bridge version 1.0.41
      Version 30.0.4-6686687
      API Levels: 23, 26, 27, 28, 29, 30
      Build Tools: 28.0.2, 28.0.3, 29.0.0, 29.0.1, 29.0.2, 29.0.3, 30.0.0
      System Images: android-21 | Intel x86 Atom_64, android-22 | Intel x86 Atom_64, android-23 | Google APIs Intel x86 Atom_64, android-24 | ARM 64 v8a, android-25 | Intel x86 Atom_64, android-25 | Google APIs ARM 64 v8a, android-26 | Intel x86 Atom_64, android-27 | Intel x86 Atom_64, android-28 | Intel x86 Atom, android-28 | Intel x86 Atom_64, android-28 | Google APIs Intel x86 Atom_64, android-28 | Google X86_ARM Intel x86 Atom, android-29 | Intel x86 Atom_64, android-29 | Google APIs Intel x86 Atom, android-29 | Google Play Intel x86 Atom_64
      Android NDK: Not Found
  IDEs:
    Android Studio: 4.1 AI-201.8743.12.41.6858069
    Xcode: 12.0.1/12A7300 - /usr/bin/xcodebuild
  Languages:
    Java: 1.8.0_201 - /usr/bin/javac
    Python: 2.7.16 - /usr/bin/python
  npmPackages:
    @react-native-community/cli: Not Found
    react: 16.13.1 => 16.13.1 
    react-native: 0.63.2 => 0.63.2 
  npmGlobalPackages:
    *react-native*: Not Found
```

**Comment by mrbrentkelly** [comment]
I believe what I'm experiencing is similar to https://github.com/wix/Detox/issues/1427 but I don't see a resolution on that issue, any ideas on what I could be doing wrong?

**Comment by mrbrentkelly** [comment]
Is there any advice on how to resolve this issue? If I have an Android Emulator running before I execute `detox test` then I get the error

```
detox[42581] ERROR: [EmulatorTelnet.js/TELNET_TIMEOUT] 
detox[42581] ERROR: Error: response not received
```

Also if I set `shutdownDevice` to `true` in my `detoxrc` then I get the above error after my test run completes.

So I have to shutdown my emulator before and after each test run.

This used to work just fine so I'm not sure what's up? I'm on Detox version `17.11.1`. Any help would be greatly appreciated.

**Comment by mrbrentkelly** [comment]
Ok so I've dug into the Detox source code and turns out connecting via telnet was working just fine and it was actually this line that was failing...

https://github.com/wix/Detox/blob/3b47b0cef8c62205cff1e6711342f7140baa254a/detox/src/devices/drivers/android/tools/EmulatorTelnet.js#L29

So wondering why the `auth` call was failing I noticed that the `auth` variable was empty. This lead me to the line before which pulls out an auth key from a file on your home directory called `.emulator_console_auth_token`.

https://github.com/wix/Detox/blob/3b47b0cef8c62205cff1e6711342f7140baa254a/detox/src/devices/drivers/android/tools/EmulatorTelnet.js#L28

So I open the file and its empty 🤔. So I just delete the file, re-run my tests and everything works perfectly, mystery solved 🎉. Hopefully this helps any one else running into the problem.

Maybe the Detox code should be deleting this file if it exists before attempting the telnet (to help mitigate this issue)? 

**TL;DR: Deleting the `.emulator_console_auth_token` file on my home directory fixed the issue for me.**

**Comment by Quandm1195** [comment]
Comment here so it will go up to google search. The issue is annoying since it happens out of nowhere. Thank you for the solution.
Keywords: EmulatorTelnet.js/TELNET_TIMEOUT

**Comment by WilGatlin** [comment]
Thanks @mrbrentkelly for the solution. Was this ever addressed in an update?

**Comment by d4vidi** [comment]
We might be removing telnet altogether from our project, soon. @noomorph could you specify the issue# to track?
